compiler/noirc_frontend/src/elaborator/expressions.rs

```diff
@@ -1220,7 +1220,20 @@ impl<'context> Elaborator<'context> {
         });
 
         let captured_vars = vecmap(&lambda_context.captures, |capture| {
-            self.interner.definition_type(capture.ident.id)
+            let typ = self.interner.definition_type(capture.ident.id);
+            if let Some(definition) = self.interner.try_definition(capture.ident.id) {
+                if definition.mutable && !typ.is_mutable_ref() {
+                    let location = definition.location;
+                    self.push_err(
+                        TypeCheckError::MutableCaptureWithoutRef {
+                            name: definition.name.clone(),
+                            span: location.span,
+                        },
+                        location.file,
+                    );
```

Comment:
Is it necessary to issue an error here? The original intent for lambda captures was not to prevent capturing a mutable variable completely, but rather that any mutable variables should be captured by value and thus be immutable within the lambda. E.g. the equivalent of `let mut x = Foo();   bar(x)` where `fn bar(y: Foo) { ... }`

Reply:
It is not. This is what we currently allow. Taking the example from https://github.com/noir-lang/noir/issues/4795#issuecomment-2086646284:
```noir
fn main() {
    let mut x = 3;

    let f = || {
        x += 2;
    };

    f();
    assert(x == 5);
}
```
This example will fail on master. `assert(x == 3);` passes.

Reply:
I thought we desired to have an error, but I could just have this be a warning.

Reply:
If we wanted to capture a mutable variable immutably it would require devs to reassign their variable before capturing in a lambda. This is a bit cumbersome in my opinion.

Reply:
Ah I see. I misunderstood as I thought from the original issue you desired this restrictive of a check. I'll update and add this test.